### DISCUSSION TITLE
`ITestOutputHelper` for Collection Fixture?

### DISCUSSION THREAD

**Discussion by mu88** [xunit/xunit#2391] [Question] [answered]
Would it be possible to provide an instance of `ITestOutputHelper` to Collection Fixtures?

I want to write output to the test runner's UI. That is not possible at the moment using `IMessageSink`.

**Comment by bradwilson** [comment, marked as answer]
Given the existing design, it isn't possible, because `ITestOutputHelper` is scoped to the execution of a single test.

Writing diagnostic messages to the diagnostic message sink does show in most runners if you have [enabled viewing diagnostic messages](https://xunit.net/docs/configuration-files#diagnosticMessages) (this is turned off by default).

**Reply by mu88** [reply]
Hmmm 🤔 even with the following `xunit.runner.json`, the output is not shown in the test runner's output window:
```json
{
  "$schema": "https://xunit.net/schema/current/xunit.runner.schema.json",
  "diagnosticMessages": true,
  "internalDiagnosticMessages": true
}
```

Rider:
![image](https://user-images.githubusercontent.com/4560672/137680723-cd2a85c8-9ead-4a2d-8519-22a969161c3f.png)

VS plain:
![image](https://user-images.githubusercontent.com/4560672/137680755-e6156ee3-48ba-4eea-8273-c47e38ca91a3.png)

VS using ReSharper:
![image](https://user-images.githubusercontent.com/4560672/137680776-9f7ccc16-e0e1-46df-b705-27a413786e5b.png)

However, I see the log output in the corresponding log file.

I've taken a look at the code and found a couple other implementations of `IMessageSinkMessage`, i. a. `TestOutput`. `ITestOutputHelper` is using this class as well. The only thing that stops me from using this class is the `ITest` constructor parameter. Can I somehow obtain this instance, e. g. with the help of a public static helper class?

**Reply by mu88** [reply]
@bradwilson Ping

**Reply by mu88** [reply]
Would it be sufficient changing [this block](https://github.com/xunit/xunit/blob/v2/src/xunit.execution/Sdk/Frameworks/Runners/XunitTestClassRunner.cs#L76) from
```cs
                object arg;
                if (p.ParameterType == typeof(IMessageSink))
                    arg = DiagnosticMessageSink;
                else
                if (!collectionFixtureMappings.TryGetValue(p.ParameterType, out arg))
                    missingParameters.Add(p);
                return arg;
```
to
```cs
                object arg;
                if (p.ParameterType == typeof(IMessageSink))
                    arg = DiagnosticMessageSink;
                else if (p.ParameterType == typeof(ITestOutputHelper))
                    arg = new TestOutputHelper();
                else
                if (!collectionFixtureMappings.TryGetValue(p.ParameterType, out arg))
                    missingParameters.Add(p);
                return arg;
```
?

**Reply by bradwilson** [reply]
No, that isn't a fix, because the `TestOutputHelper` requires an active test.

Diagnostic messages inside Visual Studio show up in the Tests tab of the output window:

![image](https://user-images.githubusercontent.com/16944/140621472-7320e5e0-9e61-47b5-8ea7-e959e23c1a75.png)

They will show up on the console in `dotnet test`:

![image](https://user-images.githubusercontent.com/16944/140621488-e7662752-5837-43a5-bd0a-16141b1208ad.png)

Note that we do not control where these messages show up. That UI is controlled by Visual Studio and `dotnet test`.

If you're still not seeing the output, make sure you see your xunit.runner.json in the output folder (which requires setting the `ItemGroup` as mentioned in [https://xunit.net/docs/configuration-files](https://xunit.net/docs/configuration-files)):

![image](https://user-images.githubusercontent.com/16944/140621539-dd99f6d3-d644-467d-9ab4-c08119347a38.png)

**Reply by mu88** [reply]
I know that the messages appear within the _Output_ window and the log file - but I want to see them [here](https://user-images.githubusercontent.com/4560672/137680755-e6156ee3-48ba-4eea-8273-c47e38ca91a3.png) and [here](https://user-images.githubusercontent.com/4560672/137680776-9f7ccc16-e0e1-46df-b705-27a413786e5b.png), like with `ITestOutputHelper`. If you can show an approximate way to what a solution might look like, I'm happy to implement it.

**Reply by bradwilson** [reply]
I can't help you with that. I don't control what shows up in those windows, and `ITestOutputHelper` is not available to anything except tests.

**Reply by mu88** [reply]
So your recommendation is to not use Class and Collection Fixtures at all when being dependant on `ITestOutputHelper`?

**Reply by bradwilson** [reply]
I'm sorry, I don't understand the question, but I'll try to re-state my answer: We can't control how Visual Studio or Resharper shows those messages, because we don't control those UIs. I'm not sure what class or collection fixtures has to do with it.

**Reply by mu88** [reply]
I'm sorry I gave you so little context - now a bit more in detail: originally we were using Class Fixtures to group tests. In addition, we were using `ITestOutputHelper` and the log output was nicely shown within the mentioned output windows.
Due to technical reasons we had to switch to Collection Fixtures and therefore to `IMessageSink`. Now with Collection Fixtures we cannot longer use `ITestOutputHelper` since that throws a runtime exception.

So our basic demand is: use Collection Fixtures and display the log output within the Output windows.

Don't get me wrong, I can perfectly live with the fact that `ITestOutputHelper` is not compatible to Collection Fixtures. I just would like to understand why I can use `ITestOutputHelper` together with Class Fixtures but not with Collection Fixtures - because my colleagues will ask me the same 😉

**Reply by bradwilson** [reply]
`ITestOutputHelper` isn't compatible with any kind of fixture (class or collection). It's possible you're mis-using it in a way that accidentally works. The instance of the test output helper is strongly tied to one specific test, and will fail if you try to use it outside the context of that test.

**Reply by bradwilson** [reply]
That said, you can take an instance of `ITestOutputHelper` that's injected into your test class's constructor, and pass it to any kind of fixture you want, so long as you make sure the fixture never uses the output helper outside of its normal lifetime (you'll know if you do, because it'll throw).

**Reply by mu88** [reply]
Got it, understood. So we were actually misusing it - thanks for clarifying this.

**Reply by fakhrulhilal** [reply]
I have the same problem as @mu88. Basically what I want is showing message in test output window. I use this to redirect logging (instead of writing to file, etc). So it's good for TDD. It's possible with [NUnit](https://docs.nunit.org/articles/vs-test-adapter/Trace-and-Debug.html) and still goes to each test output. It would be nice to support this feature.
